libp2p-xtra/src/node.rs

```diff
@@ -0,0 +1,508 @@
+use crate::multiaddress_ext::MultiaddrExt as _;
+use crate::upgrade;
+use crate::Connection;
+use crate::Substream;
+use anyhow::bail;
+use anyhow::Context as _;
+use anyhow::Result;
+use async_trait::async_trait;
+use futures::future::BoxFuture;
+use futures::stream::BoxStream;
+use futures::AsyncRead;
+use futures::AsyncWrite;
+use futures::TryStreamExt;
+use libp2p_core::identity::Keypair;
+use libp2p_core::transport::Boxed;
+use libp2p_core::transport::ListenerEvent;
+use libp2p_core::Multiaddr;
+use libp2p_core::PeerId;
+use libp2p_core::Transport;
+use multistream_select::NegotiationError;
+use multistream_select::Version;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::marker::PhantomData;
+use std::time::Duration;
+use thiserror::Error;
+use tokio_tasks::Tasks;
+use xtra::message_channel::StrongMessageChannel;
+use xtra_productivity::xtra_productivity;
+use xtras::SendAsyncSafe;
+
+/// An actor for managing multiplexed connections over a given transport.
+///
+/// The actor does not inflict any policy on connection and/or protocol management.
+/// New connections can be established by sending a [`Connect`] messages. Existing connections can
+/// be disconnected by sending [`Disconnect`]. Listening for incoming connections is done by sending
+/// a [`ListenOn`] message. To list the current state, send the [`GetConnectionStats`] message.
+///
+/// The combination of the above should make it possible to implement a fairly large number of
+/// policies. For example, to maintain a connection to a specific node, you can regularly check if
+/// the connection is still established by sending [`GetConnectionStats`] and react accordingly
+/// (f.e. sending [`Connect`] in case the connection has disappeared).
+///
+/// Once a connection with a peer is established, both sides can open substreams on top of the
+/// connection. Any incoming substream will - assuming the protocol is supported by the node -
+/// trigger a [`NewInboundSubstream`] message to the actor provided in the constructor.
+/// Opening a new substream can be achieved by sending the [`OpenSubstream`] message.
+pub struct Node {
+    transport: Boxed<Connection>,
+    tasks: Tasks,
+    controls: HashMap<PeerId, (yamux::Control, Tasks)>,
+    inbound_substream_channels:
+        HashMap<&'static str, Box<dyn StrongMessageChannel<NewInboundSubstream>>>,
+    listen_addresses: HashSet<Multiaddr>,
+    inflight_connections: HashSet<PeerId>,
+    connection_timeout: Duration,
+}
+
+/// Open a substream to the provided peer.
+///
+/// Fails if we are not connected to the peer or the peer does not support any of the requested
+/// protocols.
+#[derive(Debug)]
+pub struct OpenSubstream<P> {
+    peer: PeerId,
+    protocols: Vec<&'static str>,
+    marker_num_protocols: PhantomData<P>,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum Single {}
+
+#[derive(Clone, Copy, Debug)]
+pub enum Multiple {}
```

Comment:
❓ I would expect these to be `struct`s. Any reason why that is not the case?

Reply:
Because they should never be instantiated really. These is the poor-man's version of dependent types which Rust does not have (yet) :D

Ideally, I'd directly tie the return value of the `OpenSubstream` message to the length of `protocols` array: If only 1 protocol, then I don't need to return the negotiated protocol, if more protocols, I need to say which one was negotiated.